File: android/app/src/main/kotlin/com/example/nutrilens_mobile_app/ar/ArController.kt
package com.example.nutrilens_mobile_app.ar

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.google.ar.core.ArCoreApk
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Bridges the Flutter `nutrilens/ar` MethodChannel + `nutrilens/ar/distance`
 * EventChannel to the live [ArPlatformView]. Handles capability probing,
 * realtime distance streaming and still-frame capture.
 */
class ArController(
    private val context: Context,
    messenger: BinaryMessenger,
) {
    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null
    private var view: ArPlatformView? = null

    init {
        MethodChannel(messenger, "nutrilens/ar").setMethodCallHandler { call, result ->
            onMethodCall(call, result)
        }
        EventChannel(messenger, "nutrilens/ar/distance").setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                    eventSink = events
                }

                override fun onCancel(arguments: Any?) {
                    eventSink = null
                }
            },
        )
    }

    fun attachView(view: ArPlatformView) {
        this.view = view
    }

    /** No-ops if no AR view is currently active — safe to call unconditionally
     *  from Activity.onPause/onResume regardless of which screen is showing. */
    fun pauseActiveSession() {
        view?.pauseSession()
    }

    fun resumeActiveSession() {
        view?.resumeSession()
    }

    fun emitDistance(distanceCm: Double?, stable: Boolean, depthSource: String) {
        val sink = eventSink ?: return
        mainHandler.post {
            sink.success(
                buildMap<String, Any> {
                    if (distanceCm != null) put("distance", distanceCm)
                    put("stable", stable)
                    put("depthSource", depthSource)
                },
            )
        }
    }

    /**
     * Reports that the ARCore session itself failed to start (e.g. camera
     * permission denied at the OS level despite the Dart-side pre-check, or
     * the camera being held by another app). Without this, [ensureSession]
     * failures previously degraded silently into "distance stays null
     * forever", indistinguishable on screen from "still searching for a
     * surface".
     */
    fun emitSessionError(message: String) {
        val sink = eventSink ?: return
        mainHandler.post {
            sink.success(
                buildMap<String, Any> {
                    put("stable", false)
                    put("depthSource", "none")
                    put("error", message)
                },
            )
        }
    }

    private fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "checkCapability" -> result.success(checkCapability())
            "requestInstall" -> {
                requestInstall()
                result.success(null)
            }
            "captureFrame" -> captureFrame(result)
            "dispose" -> {
                view = null
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    /**
     * Launches the "Google Play Services for AR" install/update flow for a
     * [ArCapability.needsInstall] device. Per ARCore's own contract this must
     * only be called in direct response to an explicit user action (the
     * Dart-side "Cài đặt" button tap satisfies that). The result isn't
     * synchronous — the host Activity pauses while the user is in the Play
     * Store, and the Dart side re-probes `checkCapability` once the app
     * resumes, so there's nothing useful to return here beyond "we tried".
     */
    private fun requestInstall() {
        val activity = context as? Activity ?: return
        try {
            ArCoreApk.getInstance().requestInstall(activity, true)
        } catch (e: Exception) {
            // UnavailableDeviceNotCompatibleException / UnavailableUserDeclinedInstallationException:
            // nothing actionable here — the resume-triggered re-check on the
            // Dart side falls back to the plain camera regardless.
        }
    }

    private fun checkCapability(): String {
        return try {
            when (ArCoreApk.getInstance().checkAvailability(context)) {
                ArCoreApk.Availability.SUPPORTED_INSTALLED -> "supported"
                ArCoreApk.Availability.SUPPORTED_APK_TOO_OLD,
                ArCoreApk.Availability.SUPPORTED_NOT_INSTALLED -> "needsInstall"
                ArCoreApk.Availability.UNKNOWN_CHECKING -> "needsInstall"
                else -> "unsupported"
            }
        } catch (e: Exception) {
            "unsupported"
        }
    }

    private fun captureFrame(result: MethodChannel.Result) {
        val target = view
        if (target == null) {
            result.error("no_view", "AR view is not active.", null)
            return
        }
        // capture() blocks awaiting the GL thread, so run it off the main thread.
        Thread {
            val captured = target.capture()
            mainHandler.post {
                if (captured == null) {
                    result.error("capture_failed", "Could not capture AR frame.", null)
                } else {
                    result.success(captured)
                }
            }
        }.start()
    }
}

File: ios/Runner/ArController.swift
import ARKit
import Flutter
import Foundation

/// Bridges the Flutter `nutrilens/ar` MethodChannel + `nutrilens/ar/distance`
/// EventChannel to the live `ArKitPlatformView`.
final class ArController: NSObject, FlutterStreamHandler {

    private var eventSink: FlutterEventSink?
    private weak var view: ArKitPlatformView?

    init(messenger: FlutterBinaryMessenger) {
        super.init()

        let methodChannel = FlutterMethodChannel(name: "nutrilens/ar", binaryMessenger: messenger)
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }

        let eventChannel = FlutterEventChannel(name: "nutrilens/ar/distance", binaryMessenger: messenger)
        eventChannel.setStreamHandler(self)
    }

    func attach(_ view: ArKitPlatformView) {
        self.view = view
    }

    /// No-ops if no AR view is currently active — safe to call unconditionally
    /// from scene lifecycle callbacks regardless of which screen is showing.
    func pauseActiveSession() {
        view?.pauseSession()
    }

    func resumeActiveSession() {
        view?.resumeSession()
    }

    func emitDistance(_ distanceCm: Double?, stable: Bool, depthSource: String) {
        guard let sink = eventSink else { return }
        var payload: [String: Any] = ["stable": stable, "depthSource": depthSource]
        if let distanceCm = distanceCm {
            payload["distance"] = distanceCm
        }
        DispatchQueue.main.async { sink(payload) }
    }

    /// Reports that the ARKit session itself failed (`ARSessionDelegate.session(_:didFailWithError:)`),
    /// e.g. camera permission revoked at the OS level despite the Dart-side
    /// pre-check. Without this, a session failure was previously dropped
    /// entirely — the UI kept showing "still searching for a surface" with
    /// no way to tell the user was stuck for a reason that camera-shake
    /// would never fix.
    func emitSessionError(_ message: String) {
        guard let sink = eventSink else { return }
        let payload: [String: Any] = ["stable": false, "depthSource": "none", "error": message]
        DispatchQueue.main.async { sink(payload) }
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    // MARK: - Method handling

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "checkCapability":
            result(ARWorldTrackingConfiguration.isSupported ? "supported" : "unsupported")
        case "captureFrame":
            // ARSCNView must be accessed on the main thread.
            DispatchQueue.main.async { [weak self] in
                guard let captured = self?.view?.capture() else {
                    result(FlutterError(code: "capture_failed", message: "Could not capture AR frame.", details: nil))
                    return
                }
                result(captured)
            }
        case "dispose":
            view = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
